### ISSUE TITLE
skillopt's mini-batch focus block is always empty: train ids filtered against val rows (disjoint splits)

### ISSUE THREAD

**Issue by OsherElhadad** [skillberry-ai/cap-evolve#371]
## Summary

`skillopt` builds its per-step mini-batch from **train** ids and then passes them to `harness._focus_instructions`, which filters the parent candidate's **val** rows. Train and val are disjoint by construction, so the intersection is always empty and **every mini-batch step renders an empty focus block** — no failing-task list, no failure index, no protect-the-passing-ids block.

SkillOpt's headline mechanism is "epochs × mini-batches with a decaying textual learning rate". The decay and the epoch structure work. The mini-batch *content* does not reach the optimizer. So the thing being annealed is an instruction block that is empty.

This is the same defect that PR #370 just fixed for `hill-climb`'s `cyclic` and `hardest-first` focus modes; `skillopt` is the third instance and was left untouched there because it is a different module.

## Evidence

The mini-batch ids come from train (`core/cap_evolve/skillopt.py:230`):

```python
train_ids = list(run_dir.read_splits().train)
```

They are then handed to `_focus_instructions` together with the **val** result (`skillopt.py:291`):

```python
instructions = harness._focus_instructions(current_val, minibatch_ids, label)
```

`_focus_instructions` filters the passed result's per-task rows by those ids — and the two id sets cannot overlap, because the splits are disjoint slices of one ordered id list (`core/cap_evolve/splits.py:117-119`):

```python
train = ids[:n_tr]
val   = ids[n_tr:n_tr + n_va]
test  = ids[n_tr + n_va:n_tr + n_va + n_te]
```

`train ∩ val = ∅`, so the filter yields nothing every time. The rendered label still reads `(mini-batch of N train tasks, L=…)`, which makes the failure look like it is working — the count is real, the content is not.

## Why it went unnoticed

The same reason PR #370 gives for the `hill-climb` instance: **the `mock` optimizer never reads the prompt.** Every deterministic test and every zero-API example run passes with an empty focus block, because nothing downstream asserts on the rendered instructions. The bug is only observable in a paid run, by its absence — the optimizer simply proposes as if it had no per-task signal, which is indistinguishable from a weak proposal.

That is the important lesson here and it generalizes: assertions on *rendered prompt content* are the only thing that catches this class. PR #370 added a parametrized e2e test asserting on the rendered prompt for hill-climb; `skillopt` needs the sibling.

## Fix

Mirror PR #370's fix:

1. Order/derive the focus ids over the **val** split, so they can actually match `current_val`'s rows. If the intent was genuinely to reflect on train tasks, then the *result* passed in must be a train-split result, not `current_val` — decide which, because the two readings imply different costs. #370 chose val for hill-climb; consistency argues for val here, and it avoids paying for a train evaluation whose only purpose is an ordering.
2. Add a test that asserts on the **rendered prompt**, not on schedule names or ids — e.g. that a step's instructions contain at least one failing task id drawn from the parent's val rows, and that the protect-passing block is non-empty when passing tasks exist. Without this, the fix can silently regress exactly as the original did.
3. Check the `hardest-first` lesson applies: #370 found that mode paid for a whole pre-loop train evaluation purely to build an ordering it could not use. Verify `skillopt` is not doing an analogous wasted evaluation; if it is, delete it.

## Acceptance criteria

- [ ] A mini-batch step's rendered instructions contain real per-task failure content from the parent's val rows.
- [ ] A test asserting on rendered prompt content, which fails if the id sets stop intersecting.
- [ ] No evaluation is paid for solely to build an ordering that is then unusable.
- [ ] `skills/algorithms/skillopt/SKILL.md` describes what the mini-batch actually contains (this is also covered by the skill review in #331 — coordinate rather than duplicate).

## Provenance

Found by PR #370 while fixing the identical bug in `hill-climb`'s `cyclic`/`hardest-first` modes, and reported as a cross-skill item it deliberately did not touch. Re-verified here directly against main: `skillopt.py:230`, `skillopt.py:291`, and the disjoint slicing at `splits.py:117-119`.

**Comment by OsherElhadad** [comment]
## Post-merge re-verification against current `main` (`af7cf9d6`), plus two more affected sites

Re-verified this issue end to end after #370 merged. It is still live, and the blast radius is **three** call sites, not one — the issue body cites only `skillopt.py:291` (now `:300`).

### 1. The normal mini-batch step — confirmed (`skillopt.py:300`)

Zero-API repro (custom adapter: 12 tasks, even `n` passes, odd `n` fails, so val is genuinely mixed):

```
val=['t3', 't7', 't4']  train=['t1', 't9', 't8', 't5', 't10', 't2']  baseline val=0.333
val per-task: [('t3', 0.0), ('t4', 1.0), ('t7', 0.0)]

rendered skillopt prompt:
  Focus: epoch 1/1 step 1/3 (mini-batch of 2 train tasks, L=4). Current val reward 0.333: 0 solid / 0 flaky / 0 failing of 0 tasks.
  ## No actionable failures in focus — seek a robustness/generalization gain that does not regress the solid tasks.
  ## Currently PASSING (1 task(s)) — your edit MUST NOT regress these
```

Note the control: the *same* run through `hill_climb_loop` with `--focus cyclic` renders `Focus: task t3 … 1 failing of 1 tasks` and `## (a) 1 ALWAYS-failing task(s)` with the real feedback line. So the difference is purely the id set, exactly as diagnosed.

One thing did change as a side effect of #370: `_focus_instructions` now classifies the protect-passing block over `per_all` rather than the focused subset, so skillopt's prompt **does** now carry a non-empty `## Currently PASSING` block. The failure index is still empty. Half-fixed by accident, which makes it slightly harder to spot.

### 2. The epoch-boundary slow/meta update — also affected (`skillopt.py:491`)

Same shape, not mentioned in the issue body:

```python
sample_ids = rng.sample(train_ids, n) ...            # skillopt.py:450-455 — TRAIN
instructions = ctx.instructions(current_val, sample_ids, f"epoch {epoch} slow/meta update", ...)
```

So the gated epoch-boundary consolidation — the step whose whole job is fixing longitudinal regressions — also renders an empty failure index.

### 3. The within-epoch failure-pattern buffer — also empty (`skillopt.py:323,327`)

```python
patterns = _failure_patterns(cand_val.per_task, focus_ids=minibatch_ids)   # val rows, train ids
"n_fail": _n_fail(cand_val.per_task, focus_ids=minibatch_ids),
```

```
$ python -c "from cap_evolve.skillopt import _failure_patterns, _n_fail; ..."
patterns: []
n_fail  : 0
(control, focus_ids=['t3','t7']) patterns: [{'pattern': 'wrong result', 'task_ids': ['t3','t7'], 'n': 2}]
```

So the "within-epoch rejected-edit **+ failure-pattern** buffer" that SKILL.md advertises as a headline mechanism carries `failure_patterns: []` and `n_fail: 0` on every step of every run. Two of skillopt's three distinctive mechanisms (the mini-batch focus and the failure-pattern buffer) are inert; only the LR schedule works.

### Note on the fix

The issue's fix option 1 ("derive the focus ids over val") is the right one, but it is **not** a one-line swap: `steps_per_epoch` is `ceil(n_train / group)` (`skillopt.py:249`) and drives `total_steps` → the whole textual-LR schedule. Switching the id source to val changes the step count, the schedule, and budget consumption. Either accept that (val is typically smaller, so fewer steps per epoch) or decouple *pacing* (keep it over train counts) from the *focus set* (val ids) — worth deciding explicitly in the PR rather than discovering it in test churn.

Acceptance criteria should also cover sites 2 and 3, and the rendered-prompt test should cover the slow-update step, not just a normal mini-batch step.
